B2: Dynamically adjust --b2-chunk-size when necessary

Slugger opened the pull request:
<!--
Thank you very much for contributing code or documentation to rclone! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

#### What is the purpose of this change?
<!--
Describe the changes here
-->
Add support for dynamic modification of `--b2-chunk-size` when necessary as discussed in #4643.

Created a new package, `chunksize` that implements a chunk size calculator, which is then called by B2 to dynamically adjust the chunk size when needed.  This package was created with the intent to eventually leverage this calculator in all backends that need to do dynamic chunk size adjustments (S3 at a minimum).

#### Was the change discussed in an issue or in the forum before?

<!--
Link issues and relevant forum posts here.
-->
Yes.  #4643
#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

ncw commented on fs/dynchunksize/dynchunksize.go:
```diff
@@ -0,0 +1,37 @@
+// Dynamically calculate suitable chunk size for large uploads
+package dynchunksize
```
I'm not a fan of this name!

Perhaps call the package `chunksize` and the function `Calculator`?

ncw commented on fs/dynchunksize/dynchunksize.go:
```diff
@@ -0,0 +1,37 @@
+	"github.com/rclone/rclone/fs"
+)
+
+const oneMiB = int64(1024 * 1024)
```
There is a constant for this already in `fs.Mebi` which has the correct type.

ncw commented on fs/dynchunksize/dynchunksize.go:
```diff
@@ -0,0 +1,37 @@
+func ComputeChunkSize(fileSize fs.SizeSuffix, maxParts int, defaultChunkSize fs.SizeSuffix) fs.SizeSuffix {
+	requiredChunks := fileSize / defaultChunkSize
+	if requiredChunks < fs.SizeSuffix(maxParts) || (requiredChunks == fs.SizeSuffix(maxParts) && fileSize%defaultChunkSize == 0) {
+		fs.Debugf(nil, "size: %v, parts: %v, default: %v; returned default value", fileSize, maxParts, defaultChunkSize)
```
We should probably comment out this debug otherwise it will appear for every chunked upload.

ncw commented on fs/dynchunksize/dynchunksize.go:
```diff
@@ -0,0 +1,37 @@
+	remainder := minChunk % fs.SizeSuffix(oneMiB)
+	if remainder != 0 {
+		minChunk += fs.SizeSuffix(oneMiB) - remainder
+	} else if int64(fileSize/minChunk) == int64(maxParts) && int64(fileSize)%int64(maxParts) != 0 { // when right on the boundary, we need to add a MiB
```
I think you can probably get rid of most of the `int64()` casts in this line, as everything is `fs.SizeSuffix`.

I'm not convinced the `else if` is correct here... I think it should be (and would be safe if it was)

```go
    if remainder != 0 {
        minChunk += fs.SizeSuffix(oneMiB) - remainder
    }
    if int64(fileSize/minChunk) == int64(maxParts) && ...{
    }
```

ncw commented on fs/dynchunksize/dynchunksize.go:
```diff
@@ -0,0 +1,37 @@
+		minChunk += fs.SizeSuffix(oneMiB)
+	}
+
+	fs.Debugf(nil, "size: %v, parts: %v, default: %v, new: %v; default chunk size insufficient, returned new chunk size", fileSize, maxParts, defaultChunkSize, minChunk)
```
Not having the object to log here will make these logs not very useful. So either pass the object to log in, or comment out the Debug.

ncw reviewed (commented):
Looks like a nice change :-)

I put some comments inline.

In order to accept the new package I'd like you to add a commit which uses it for `s3` and another ideally for `azureblob` - I don't like abstractions which don't get used more than once :-)

I'll run the CI on this too so you can see if it passes or not.

Slugger commented:
Thanks for the feedback. I've addressed all the comments except the implementation in additional backends.  I'll do that next time I can put some time in -- some evening hopefully later this week.

Just dipping my toes into the project here with this relatively simple feature request I found in the backlog. Got to say, so far so good.  It's always nice to get a super quick, constructive response when starting out in a new project. Looking forward to contributing (I have a big idea, which is what brought me here, but learning to crawl before I try to climb Everest! 😄 ).

Slugger commented:
Refactored `s3` and `azureblob` to leverage the new `chunksize.Calculator()`

ncw reviewed (commented):
Changes look great - thank you :-)

I'll run the CI now.

Can you squash into 4 commits, one for the chunksize lib, then one for each backend then I'll run the CI one more time and merge - thank you :-)

Slugger commented:
Linter issues fixed, commits squashed.  Everything should be ready to go.

ncw reviewed (approved):
Thank you - looks perfect. I'll run the CI one last time then merge :-)

ncw commented:
Thank you :-)
